### ISSUE TITLE
Make ns.corporation.research() check if the research is already researched before charging the research points.

### ISSUE THREAD

**Issue by fknietzsch** [danielyxie/bitburner#3977]
Could it be that the netscript function [ns.corporation.research(divisionName, researchName)](https://github.com/danielyxie/bitburner/blob/dev/markdown/bitburner.officeapi.research.md) doesn't check whether the research is already unlocked before charging the research points from the player (the division of the corporation)?

I wrote a script to automate corporations and it constantly spends research points on "Hi-Tech R&D Laboratory" because of a missing check if it is already researched in my code. So I wonder if it is really nice of the game to show this behaviour. I would expect the function to only charge the points when there's an effect to be expected. But calling the research function on the same target over and over again seems to try to research it again and again - leaving the division poor of points :-(

So... just asking: Is this intended? Could it be fixed?

Thanks in advance!

**Comment by phyzical** [comment]
Looking at the code it may not be checking it, though ill give it a test tonight to verify but it should be as simple as adding a `if division.researched[researchName] ` and early out if true already before charging
```
export function Research(division: IIndustry, researchName: string): void {
  const researchTree = IndustryResearchTrees[division.type];
  if (researchTree === undefined) throw new Error(`No research tree for industry '${division.type}'`);
  const allResearch = researchTree.getAllNodes();
  if (!allResearch.includes(researchName)) throw new Error(`No research named '${researchName}'`);
  const research = ResearchMap[researchName];

  if (division.sciResearch.qty < research.cost)
    throw new Error(`You do not have enough Scientific Research for ${research.name}`);
  division.sciResearch.qty -= research.cost;

  // Get the Node from the Research Tree and set its 'researched' property
  researchTree.research(researchName);
  division.researched[researchName] = true;
}

```

**Comment by phyzical** [comment]
OH also @fknietzsch this is a perfect example of when to error proof your code and check/reduce what to research what you havnt instead of trying everything over and over (means more perf for your script 😉  )

**Comment by fknietzsch** [comment]
> OH also @fknietzsch this is a perfect example of when to error proof your code and check/reduce what to research what you havnt instead of trying everything over and over (means more perf for your script 😉 )

Hi @phyzical , thanks for the reply. 
I added/corrected the check in my code after I noticed that behaviour. It wasn't intended from my side to try researching it over and over again, with my script. It was just an error in my code. But still I thought that this might not be an actually intended behaviour of the game function. So I decided to open this issue.

Maybe this can help to improve the game which I love so much an have spend hundreds of hours with already 😊  Huge thanks to all the devs and contributors!

Best regards
